### ISSUE TITLE
LOCALE_ID is being set before APP_INITIALIZER runs

### ISSUE THREAD

**Issue by elias-garcia** [angular/angular#31465]
# 🐞 bug report

### Affected Package
The issue is caused by package `@angular/core`.

### Is this a regression?

Yes, the previous version in which this bug was not present was: `7.2.15`

### Description

Angular is trying to set the `LOCALE_ID` before the `APP_INITIALIZER` runs to set the runtime i18n value for Ivy. This does not allow the application to fetch the locale from a backend. This is happening without Ivy mode activated.

The screenshot below is from file: `node_modules/@angular/core/fesm2015/core.js`.

![Captura de pantalla de 2019-07-09 13-48-44](https://user-images.githubusercontent.com/6737805/60885940-04515300-a251-11e9-8063-2c5e9123bae2.png)

## 🌍  Your Environment

**Angular Version:**
```
Angular CLI: 8.1.0
Node: 10.16.0
OS: linux x64
Angular: 8.1.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.801.0
@angular-devkit/build-angular     0.801.0
@angular-devkit/build-optimizer   0.801.0
@angular-devkit/build-webpack     0.801.0
@angular-devkit/core              8.1.0
@angular-devkit/schematics        8.1.0
@angular/cdk                      8.0.2
@angular/material                 8.0.2
@ngtools/webpack                  8.1.0
@schematics/angular               8.1.0
@schematics/update                0.801.0
rxjs                              6.5.2
typescript                        3.4.5
webpack                           4.35.2
```

**Comment by jasonaden** [comment]
Appears related to #29249

**Comment by ocombe** [comment]
Do you have an example where this is an issue for view engine?
`setLocaleId` is only used for ivy code, and this piece of code is only there temporarily to allow runtime i18n (for ivy) to work with the current view engine code until we have the new i18n service available.

**Comment by elias-garcia** [comment]
Hi @ocombe! That code is intended to being used by Ivy, but it's not. We've only updated Angular version from 7.2.15 to 8.1 and the app our app was broken (we haven't enabled any Ivy flag). We are setting the `LOCALE_ID` with a property `currentLocale` from a `LocaleService`. That property is being set from a backend call using the `APP_INITIALIZER` token, so the application does not even start, because Angular is trying to set the `LOCALE_ID` before we get the current locale for the user from the backend.

**Comment by ocombe** [comment]
Is there an error or something? Without a repro it's hard to figure out what's going wrong.

**Comment by jasedwards** [comment]
I think [this](https://stackblitz.com/edit/angular-nkerqf) reproduces it.  the localeId is used before APP_INITIALIZER runs

**Comment by elias-garcia** [comment]
As @jasedwards said, the project was using Angular 7.x.x... So i cannot post a repro. The error is that we are trying to set `LOCALE_ID` to the value from the service and then convert it to lower case. So, as the locale has not yet been recovered from the backend, it throws an exception `Cannot read property 'toLowerCase' of undefined`.

**Comment by ocombe** [comment]
oh I see ! Thanks I'll fix that

**Comment by botondus** [comment]
We ran into the same issue after upgrading to Angular 8 (and without Ivy enabled). We set the LOCALE_ID based on a backend API call (through a `Resolver`), and now the application tries to set the LOCALE_ID before the backend API call is done, so it is set to undefined and causes the `Cannot read property 'toLowerCase' of undefined.` error

(I've posted it here: https://stackoverflow.com/questions/57115654/lazy-loading-and-setting-locale-id-at-runtime-from-an-api-in-angular-8-applicati)

Looking forward to a fix for this.

**Comment by botondus** [comment]
Hi @ocombe , happy to see that you've developed a fix for this. Is there any ETA on when this will land in a new Angular patch release? Thanks! (I see it's not in 8.1.3)

**Comment by ocombe** [comment]
ah that's a mistake if it wasn't. I asked @AndrewKushnir to check this and to add it to the next patch release

**Comment by amiram** [comment]
@botondus Looks like it is working well on 8.2.0

**Comment by rqx110** [comment]
@ocombe it looks like this bug still exist on 8.2.0.

I just use this https://stackblitz.com/edit/angular-nkerqf, and upgrade angular to 8.2, the error still exist.

**Comment by skovalyov** [comment]
> @botondus Looks like it is working well on 8.2.0

It depends on the exact situation. At least the problem mentioned by @jasedwards with example at https://stackblitz.com/edit/angular-nkerqf persists. I forked his example and upgraded dependencies at https://stackblitz.com/edit/angular-y6fafg and it still fails. Function `getLocale()` to initialize `LOCALE_ID` gets called immediately.

**Comment by botondus** [comment]
As far as I can see this fix/change was not included in 8.2.0 either yet @amiram @skovalyov

**Comment by amiram** [comment]
There may be more than one issue here. I didn't use use and locale functionality, just a bit of my code executed in app initialize, and still got the same error. If you use the locale things like in the example, there is still a problem.

**Comment by ocombe** [comment]
Yes it was only included in 9.0.0, not the current 8 branch. I need to make
another PR against that branch so that it can be in the release next week.

**Comment by jasedwards** [comment]
Do I need to open up a new issue for the toLowerCase error thrown in my stackblitz?

**Comment by ocombe** [comment]
No it will be fixed by my PR.

>

**Comment by htonheim** [comment]
I can confirm that 8.2.0 is still broken, but in 9.0.0-Next this is working again. So it should all be fine again once the pull request with the fix that is in Angular 9 get's merged into 8.2 (8.2.1?). 👍 

Also, it was working (for us) all through the 8.0.x versions and broke in 8.1.x. For anyone wanting to upgrade to Angular 8, it should therefore be possible to upgrade to 8.0.3 without running in to this issue (or just wait for 8.2.1).

**Comment by botondus** [comment]
I've also tried 9.0.0-Next and can confirm that it is working fine in our use case with this fix.
